src/Modal/ModalContents.js

```diff
@@ -80,144 +88,43 @@ class ModalContents extends Util.mixin(BindMixin, KeyDownMixin) {
   }
 
   handleWindowResize() {
-    // Render the modal again if the window resizes because the height
-    // of the viewport may change and we need to adapt to that.
-    if (this.props.open) {
-      this.forceUpdate();
-    }
-  }
-
-  resetHeight() {
-    this.heightInfo = Util.clone(DEFAULT_HEIGHT);
-  }
-
-  checkContentHeightChange(prevHeightInfo, heightInfo) {
-    let {contentHeight, height, maxHeight, innerContentHeight} = heightInfo;
-    let prevInnerContentHeight = prevHeightInfo.innerContentHeight;
-    let prevHeight = prevHeightInfo.height;
-    let prevContentHeight = prevHeightInfo.contentHeight;
-    let prevMaxHeight = prevHeightInfo.maxHeight;
-
-    // Default update height to whether there is some change in height
-    let update = prevHeight !== height || prevContentHeight !== contentHeight;
-
-    if (prevInnerContentHeight != null) {
-      // Make sure to update heightInfo with new height difference
-      let difference = innerContentHeight - prevInnerContentHeight;
-      if (difference !== 0 && height + difference < maxHeight) {
-        // Changes this.heightInfo as this is a reference
-        heightInfo.height += difference;
-        heightInfo.contentHeight += difference;
-        update = true;
-      }
-
-      // Make sure to update heightInfo with new max height difference
-      let maxHeightDifference = maxHeight - prevMaxHeight;
-      if (maxHeightDifference > 0 &&
-        (heightInfo.contentHeight + difference < innerContentHeight)) {
-        // Changes this.heightInfo as this is a reference
-        heightInfo.height += maxHeightDifference;
-        heightInfo.contentHeight += maxHeightDifference;
-        update = true;
+    let {props, state} = this;
+
+    // Recalculate the height of the modal if it's open and using Gemini.
+    if (props.open && props.useGemini) {
+      let viewportHeight = DOMUtil.getViewportHeight();
+
+      // If the height of the viewport is getting shorter, or if it's growing
+      // while the height is currently constrained, then we reset the restrained
+      // height to null which will cause the height to be recalculated on the
+      // next render.
+      if (viewportHeight < this.lastViewportHeight
+        || (viewportHeight > this.lastViewportHeight
+          && state.height !== null)) {
+        this.setState({height: null});
```

Comment:
Will it be OK if this `setState becomes async? (just looking at the line below)
Should we do this instead?
```
this.setState({height: null}, () => {
  this.lastViewportHeight = viewportHeight;
});
```